### PR TITLE
refactor: clarify input types for user overloads

### LINKED ISSUE #1631: Make all user-written behavior overloads consistent in what arguments they accept and what they are expected to return
### Version of Awkward Array

HEAD

### Description and code to reproduce

Probably this: arguments passed to the user-written behavior are `ak.Array`s (i.e. high-level), and it doesn't matter whether they return layouts (low-level) or `ak.Array`s.

This includes `ak.transform`.

Probably does not include Numba overloads.

### PR DESCRIPTION
**Pull request by agoose77**
Closes #1631 

----
:books: The documentation for this PR will be available at <https://awkward-array.readthedocs.io/en/agoose77-fix-ufunc-as-layout/> once Read the Docs has finished building :hammer:

### CODE REVIEW

**Comment by agoose77** [comment]
@jpivarski  could you reflect on where this sits with you? Still to decide on `__broadcast__`. That feels more low level than `ufunc` etc.

**Comment by jpivarski** [comment]
> @jpivarski could you reflect on where this sits with you? Still to decide on `__broadcast__`. That feels more low level than `ufunc` etc.

The `__broadcast__` overload is low-level; it's used inside `broadcast_and_apply`, so that one can write the string equality operation and have it not broadcast the strings to have the same lengths in the two arrays. That's very much a tree-structure manipulation thing.

File: src/awkward/operations/ak_to_dataframe.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def to_dataframe(
               3         4.0  NaN
     """
     with ak._errors.OperationErrorContext(
-        "ak.to_layout",
+        "ak.to_dataframe",
```
Comment (jpivarski) [inline src/awkward/operations/ak_to_dataframe.py:124]: Good catch!

**Review by jpivarski — COMMENTED** [review]
I take it that this part of the documentation is no longer correct:

https://github.com/scikit-hep/awkward/blob/6c5d24b19f354ad8a0806822c2bba1fc6844a52a/src/awkward/operations/ak_transform.py#L120-L135

I'll fix the documentation.

File: src/awkward/operations/ak_to_dataframe.py (hunk 1 above)
Comment (agoose77) [reply src/awkward/operations/ak_to_dataframe.py:124]: Oh, I missed that during my pass over. Good spot.

Comment (jpivarski) [reply src/awkward/operations/ak_to_dataframe.py:124]: No, actually, you caught that!

Comment (agoose77) [reply src/awkward/operations/ak_to_dataframe.py:124]: Ah yes, I meant the documentation! Always the thing one misses :/

**Comment by agoose77** [comment]
@jpivarski happy with this? I didn't touch the numba code, because numba deals with types.

**Review by jpivarski — APPROVED** [review]
My understanding is that this PR changes `ak.transform` to now require Content instances (or None) as a return value from the function that the downstream developer writes. I don't see any other changes in how developer-provided functions get called or expect as return values—I guess we decided that all other cases should stay as they are.

I ran the Vector tests; they still work with this PR.

Also, I see that `numpytype` has been removed as a `to_layout` argument.

Based on my understanding, I approve this PR and it can be merged!

**Comment by agoose77** [comment]
> My understanding is that this PR changes `ak.transform` to now require `Content` instances (or None) as a return value from the function that the downstream developer writes. I don't see any other changes in how developer-provided functions get called or expect as return values—I guess we decided that all other cases should stay as they are.

Yes, it makes `ak.transform` less permissive in what the user can return.

It *also*:
- ensures that all `_array_ufunc_custom_cast` casts its arguments to `ak.Array` (in practice I think they always are)
- removes `numpytype` from `ak.to_layout`
- ensures that `__array_function__` overload results are converted to `ak.Array`.

File: src/awkward/operations/ak_transform.py [hunk 2]
```diff
@@ -499,46 +501,33 @@ def action(layout, **kwargs):
...
 
-            if isinstance(out, ak.highlevel.Array):
-                return (out.layout,)
+            elif isinstance(out, tuple):
+                for x in out:
+                    if not isinstance(x, ak.contents.Content):
+                        raise ak._errors.wrap_error(
+                            TypeError(
+                                f"transformation must return a Content, tuple of Contents, or None, not {type(x)}\n\n{x!r}"
+                            )
```
Comment (agoose77) [inline src/awkward/operations/ak_transform.py]:
```suggestion
                            TypeError(
                                f"transformation must return a Content, tuple of Contents, or None, not a tuple containing {type(x)}\n\n{x!r}"
                            )
```

**Comment by jpivarski** [comment]
These are all good changes; it can be merged. ("Ensure" is not a change in behavior, but important for API!)

**Comment by jpivarski** [comment]
This one would close #1631, wouldn't it?

**Pull request merged**